File: backend_cpp/src/multi_camera_detector.cpp
#include "multi_camera_detector.hpp"
#include <spdlog/spdlog.h>
#include <chrono>
#include <algorithm>
#include <future>

namespace safedetect {

MultiCameraDetector::MultiCameraDetector(const std::string& model_path)
    : is_running_(false), frame_count_(0), fps_(0.0) {
    // Load YOLO model
    try {
        net_ = cv::dnn::readNetFromONNX(model_path);
        net_.setPreferableBackend(cv::dnn::DNN_BACKEND_OPENCV);
        net_.setPreferableTarget(cv::dnn::DNN_TARGET_CPU);
        spdlog::info("YOLO model loaded from {}", model_path);
    } catch (const cv::Exception& e) {
        spdlog::error("Failed to load YOLO model: {}", e.what());
    }

    // Initialize Kafka producer
    kafka_producer_ = std::make_unique<DetectionKafkaProducer>();
    kafka_producer_->start_producer();

    // Initialize audio alert
    audio_alert_ = std::make_unique<AudioAlert>();
    if (!audio_alert_->initialize()) {
        spdlog::warn("Audio alert initialization failed");
    }

    last_time_ = std::chrono::steady_clock::now();
}

MultiCameraDetector::~MultiCameraDetector() {
    stop();
}

bool MultiCameraDetector::start_cameras() {
    spdlog::info("Starting multi-camera system...");

    std::vector<std::future<std::pair<std::string, cv::VideoCapture>>> futures;

    // Launch async tasks to open cameras in parallel
    for (const auto& [zone, config] : CAMERA_CONFIG) {
        futures.push_back(std::async(std::launch::async, [zone, &config]() -> std::pair<std::string, cv::VideoCapture> {
            spdlog::info("Starting {} (Camera ID: {})...", config.name, config.camera_id);
            cv::VideoCapture cap(config.camera_id);
            if (cap.isOpened()) {
                cap.set(cv::CAP_PROP_FRAME_WIDTH, CAMERA_WIDTH);
                cap.set(cv::CAP_PROP_FRAME_HEIGHT, CAMERA_HEIGHT);
                cap.set(cv::CAP_PROP_FPS, FPS_TARGET);
                return {zone, std::move(cap)};
            } else {
                spdlog::error("{}: Failed to open camera {}", config.name, config.camera_id);
                return {zone, cv::VideoCapture()}; // empty capture
            }
        }));
    }

    int success_count = 0;
    cameras_.clear();
    camera_status_.clear();

    // Collect results
    for (auto& fut : futures) {
        auto [zone, cap] = fut.get();
        if (cap.isOpened()) {
            cameras_[zone] = std::move(cap);
            camera_status_[zone] = "available";
            spdlog::info("{}: Connected successfully", zone);
            success_count++;
        } else {
            camera_status_[zone] = "error";
        }
    }

    spdlog::info("Camera startup complete: {}/{} cameras connected", success_count, CAMERA_CONFIG.size());
    return success_count > 0;
}

void MultiCameraDetector::stop() {
    if (!is_running_) return;

    is_running_ = false;

    // Release cameras
    for (auto& [zone, cap] : cameras_) {
        if (cap.isOpened()) {
            cap.release();
            spdlog::info("Released {} camera", zone);
        }
    }
    cameras_.clear();

    // Update status
    for (const auto& [zone, _] : CAMERA_CONFIG) {
        camera_status_[zone] = "not_connected";
    }

    // Stop Kafka producer
    if (kafka_producer_) {
        kafka_producer_->stop_producer();
    }

    cv::destroyAllWindows();
    spdlog::info("Multi-camera system stopped");
}

std::unordered_map<std::string, std::string> MultiCameraDetector::get_camera_status() const {
    std::unordered_map<std::string, std::string> status;
    for (const auto& [zone, config] : CAMERA_CONFIG) {
        auto it = camera_status_.find(zone);
        std::string status_str = (it != camera_status_.end()) ? it->second : "unknown";
        status[zone] = config.name + " (ID: " + std::to_string(config.camera_id) + ") - " +
                      (CAMERA_STATUS.count(status_str) ? CAMERA_STATUS.at(status_str) : status_str);
    }
    return status;
}

std::vector<Detection> MultiCameraDetector::process_all_cameras() {
    std::vector<Detection> all_detections;

    for (auto& [zone, cap] : cameras_) {
        cv::Mat frame;
        if (!cap.read(frame)) {
            spdlog::warn("Failed to read frame from {} camera", zone);
            continue;
        }

        // Prepare frame for DNN
        cv::Mat blob = cv::dnn::blobFromImage(frame, 1 / 255.0,
                                              cv::Size(640, 640),
                                              cv::Scalar(), true, false);
        net_.setInput(blob);

        // Run inference
        std::vector<cv::Mat> outputs;
        net_.forward(outputs, net_.getUnconnectedOutLayersNames());
        if (outputs.empty()) {
            spdlog::error("No outputs from YOLO");
            continue;
        }

        cv::Mat output = outputs[0];
        float* data = (float*)output.data;

        // Log output shape
        spdlog::info("YOLO output shape: {} x {} x {}", output.size[0], output.size[1], output.size[2]);

        // Log first few detections
        for (int i = 0; i < std::min(5, (int)output.size[1]); i++) {
            int offset = i * 6;
            spdlog::info("Detection {}: x={}, y={}, w={}, h={}, conf={}, cls={}",
                         i, data[offset], data[offset+1], data[offset+2], data[offset+3], data[offset+4], data[offset+5]);
        }

        // YOLOv10 output shape: (1, 300, 6) -> [x, y, w, h, conf, cls]
        int num_boxes = output.size[1];  // 300

        std::vector<cv::Rect> boxes;
        std::vector<float> confidences;
        std::vector<int> classIds;

        for (int i = 0; i < num_boxes; i++) {
            int offset = i * 6;

            float x = data[offset + 0];
            float y = data[offset + 1];
            float w = data[offset + 2];
            float h = data[offset + 3];
            float confidence = data[offset + 4];
            int class_id = static_cast<int>(data[offset + 5]);

            // Only process "person" or "car"
            if (confidence > MODEL_CONFIDENCE &&
                OBJECT_CLASSES.count(class_id) &&
                (OBJECT_CLASSES.at(class_id) == "person" || OBJECT_CLASSES.at(class_id) == "car")) {

                int x1 = static_cast<int>(x - w / 2.0f);
                int y1 = static_cast<int>(y - h / 2.0f);
                int x2 = static_cast<int>(x + w / 2.0f);
                int y2 = static_cast<int>(y + h / 2.0f);

                // Skip invalid boxes
                if (x2 <= x1 || y2 <= y1 || w <= 0 || h <= 0) continue;

                cv::Rect bbox(x1, y1, x2 - x1, y2 - y1);
                bbox &= cv::Rect(0, 0, frame.cols, frame.rows);

                // Skip if bbox becomes invalid after clipping
                if (bbox.width <= 0 || bbox.height <= 0) continue;

                // Skip small boxes to reduce false positives
                if (bbox.width * bbox.height < 10000) continue;

                boxes.push_back(bbox);
                confidences.push_back(confidence);
                classIds.push_back(class_id);
            }
        }

        // Apply Non-Maximum Suppression
        std::vector<int> indices;
        cv::dnn::NMSBoxes(boxes, confidences, MODEL_CONFIDENCE, 0.5f, indices);

        // Limit to top 1 detection per frame to reduce false positives
        if (!indices.empty()) {
            // Sort by confidence descending
            std::sort(indices.begin(), indices.end(), [&](int a, int b) {
                return confidences[a] > confidences[b];
            });
            // Take only the top 1
            indices.resize(1);
        }

        std::vector<Detection> detections;
        for (int idx : indices) {
            detections.push_back(calculate_detection(
                boxes[idx], frame.cols, frame.rows, zone, classIds[idx], confidences[idx]));
        }

        all_detections.insert(all_detections.end(), detections.begin(), detections.end());

        if (!detections.empty()) {
            kafka_producer_->send_detections(detections);
        }
    }

    // FPS calculation
    frame_count_++;
    auto current_time = std::chrono::steady_clock::now();
    auto elapsed = std::chrono::duration_cast<std::chrono::seconds>(current_time - last_time_).count();
    if (elapsed >= 1.0) {
        fps_ = frame_count_ / elapsed;
        frame_count_ = 0;
        last_time_ = current_time;
        spdlog::info("FPS: {:.1f} | Active detections: {}", fps_, all_detections.size());
    }

    // Blind spot check
    std::vector<Detection> blind_spot_detections;
    for (const auto& detection : all_detections) {
        float x_pos = detection.position.x / POSITION_SCALE.x;
        float y_pos = detection.position.y / POSITION_SCALE.y;
        if (is_in_blind_spot(x_pos, y_pos, detection.camera_zone)) {
            blind_spot_detections.push_back(detection);
        }
    }

    if (!blind_spot_detections.empty()) {
        play_alert_sound();
        spdlog::warn("BLIND SPOT ALERT! Objects detected: {}", blind_spot_detections.size());
    }

    return all_detections;
}

void MultiCameraDetector::play_alert_sound() {
    if (audio_alert_) {
        audio_alert_->play_alert();
    }
}

bool MultiCameraDetector::is_in_blind_spot(float x_center, float y_center, const std::string& zone) const {
    auto it = BLIND_SPOT_ZONES.find(zone);
    if (it == BLIND_SPOT_ZONES.end()) return false;

    const auto& zone_coords = it->second;
    return (zone_coords.x_min <= x_center && x_center <= zone_coords.x_max &&
            zone_coords.y_min <= y_center && y_center <= zone_coords.y_max);
}

Detection MultiCameraDetector::calculate_detection(const cv::Rect& bbox, int frame_width, int frame_height,
                                                   const std::string& zone, int class_id, float confidence) const {
    Detection detection;
    detection.object = OBJECT_CLASSES.at(class_id);
    detection.confidence = confidence;
    detection.bbox = {static_cast<float>(bbox.x), static_cast<float>(bbox.y),
                      static_cast<float>(bbox.x + bbox.width), static_cast<float>(bbox.y + bbox.height)};
    detection.class_id = class_id;
    detection.camera_zone = zone;
    detection.timestamp = std::chrono::duration_cast<std::chrono::milliseconds>(
        std::chrono::system_clock::now().time_since_epoch()).count() / 1000.0;

    float x_center = (bbox.x + bbox.width / 2.0f) / frame_width;
    float y_center = (bbox.y + bbox.height / 2.0f) / frame_height;

    // Debug logging
    spdlog::info("Detection bbox: x1={}, y1={}, x2={}, y2={}, frame_size={}x{}",
                 bbox.x, bbox.y, bbox.x + bbox.width, bbox.y + bbox.height, frame_width, frame_height);
    spdlog::info("Detection center: x_center={}, y_center={}", x_center, y_center);

    detection.position.x = x_center * POSITION_SCALE.x;
    detection.position.y = y_center * POSITION_SCALE.y;

    if (zone == "left") detection.position.z = 4.0f;
    else if (zone == "right") detection.position.z = -5.0f;
    else if (zone == "rear") detection.position.z = 0.0f;
    else detection.position.z = 0.0f;

    detection.position.zone = zone;

    spdlog::info("Detection position: x={}, y={}, z={}, zone={}",
                 detection.position.x, detection.position.y, detection.position.z, detection.position.zone);

    return detection;
}

} // namespace safedetect

File: backend_Python/computer_vision/multi_camera_detector.py
"""
Multi-Camera Blind Spot Detection System
Supports multiple camera feeds for comprehensive blind spot monitoring
"""

import cv2
import numpy as np
from ultralytics import YOLO
import time
import asyncio
import json
import sys
import os
import hashlib
import hmac
sys.path.append(os.path.join(os.path.dirname(__file__), '..', '..'))
from shared.config import *
import pygame
from typing import List, Dict, Tuple, Optional
import logging
from .kafka_producer import DetectionKafkaProducer

# Set up logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class MultiCameraDetector:
    def __init__(self, model_path: str = "yolov8n.pt"):
        """Initialize the multi-camera blind spot detection system"""
        self.model = YOLO(model_path)
        self.cameras = {}  # Dictionary to store multiple camera feeds
        self.is_running = False
        self.frame_count = 0
        self.fps = 0
        self.last_time = time.time()

        # Initialize pygame for audio alerts
        self.audio_enabled = False
        try:
            pygame.mixer.init()
            self.alert_sound = self._create_beep_sound()
            self.audio_enabled = True
        except Exception as e:
            logger.warning(f"Audio device not available, alerts will be silent: {e}")
            self.alert_sound = None

        # Detection history for smoothing
        self.detection_history = []

        # Camera status tracking
        self.camera_status = {
            "left": {"status": "not_connected", "cap": None},
            "right": {"status": "not_connected", "cap": None},
            "rear": {"status": "not_connected", "cap": None}
        }

        # Initialize Kafka producer
        self.kafka_producer = DetectionKafkaProducer()
        self.kafka_producer.start_producer()

    def _create_beep_sound(self) -> pygame.mixer.Sound:
        """Create a beep sound for alerts"""
        sample_rate = 44100
        duration = ALERT_DURATION
        frequency = ALERT_BEEP_FREQUENCY

        # Generate sine wave
        t = np.linspace(0, duration, int(sample_rate * duration), False)
        wave = 0.3 * np.sin(2 * np.pi * frequency * t)

        # Convert to 16-bit PCM
        wave = (wave * 32767).astype(np.int16)

        # Create pygame sound object
        return pygame.mixer.Sound(wave.tobytes())

    def is_in_blind_spot(self, x_center: float, y_center: float, zone: str) -> bool:
        """Check if an object is in a blind spot zone"""
        zone_coords = BLIND_SPOT_ZONES[zone]
        return (zone_coords["x_min"] <= x_center <= zone_coords["x_max"] and
                zone_coords["y_min"] <= y_center <= zone_coords["y_max"])

    def calculate_position(self, bbox: List[float], frame_width: int, frame_height: int, zone: str) -> Dict[str, float]:
        """Calculate relative position of detected object"""
        x1, y1, x2, y2 = bbox
        x_center = (x1 + x2) / 2 / frame_width
        y_center = (y1 + y2) / 2 / frame_height

        # Convert to world coordinates (meters)
        world_x = x_center * POSITION_SCALE["x"]
        world_y = y_center * POSITION_SCALE["y"]

        # Set Z coordinate based on camera zone for better 3D positioning
        if zone == "left":
            world_z = 4.0  # Behind truck (negative Z)
        elif zone == "right":
            world_z = -5.0  # Behind truck (negative Z)
        elif zone == "rear":
            world_z = 0  # Further behind truck
        else:
            world_z = 0  # Default

        return {"x": world_x, "y": world_y, "z": world_z, "zone": zone}

    def start_cameras(self) -> Dict[str, bool]:
        """Start all configured cameras"""
        logger.info("🔄 Starting multi-camera system...")
        success_count = 0
        results = {}

        for zone, config in CAMERA_CONFIG.items():
            camera_id = config["camera_id"]
            try:
                logger.info(f"📹 Starting {config['name']} (Camera ID: {camera_id})...")

                cap = cv2.VideoCapture(camera_id)
                cap.set(cv2.CAP_PROP_FRAME_WIDTH, CAMERA_WIDTH)
                cap.set(cv2.CAP_PROP_FRAME_HEIGHT, CAMERA_HEIGHT)
                cap.set(cv2.CAP_PROP_FPS, FPS_TARGET)

                if cap.isOpened():
                    self.cameras[zone] = cap
                    self.camera_status[zone] = {
                        "status": "available",
                        "cap": cap,
                        "config": config
                    }
                    logger.info(f"✅ {config['name']}: Connected successfully")
                    success_count += 1
                    results[zone] = True
                else:
                    logger.warning(f"❌ {config['name']}: Failed to open camera {camera_id}")
                    self.camera_status[zone] = {
                        "status": "error",
                        "cap": None,
                        "config": config
                    }
                    results[zone] = False

            except Exception as e:
                logger.error(f"❌ {config['name']}: Error - {e}")
                self.camera_status[zone] = {
                    "status": "error",
                    "cap": None,
                    "config": config
                }
                results[zone] = False

        logger.info(f"📊 Camera startup complete: {success_count}/{len(CAMERA_CONFIG)} cameras connected")
        return results

    def get_camera_status(self) -> Dict:
        """Get status of all cameras"""
        status = {}
        for zone, info in self.camera_status.items():
            config = info["config"]
            status[zone] = {
                "name": config["name"],
                "camera_id": config["camera_id"],
                "zone": config["zone"],
                "status": info["status"],
                "description": config["description"]
            }
        return status

    async def process_all_cameras(self) -> List[Dict]:
        """Process frames from all active cameras"""
        all_detections = []

        for zone, cap in self.cameras.items():
            try:
                ret, frame = cap.read()
                if not ret:
                    logger.warning(f"⚠️  Failed to read frame from {zone} camera")
                    continue

                # Data integrity check - compute hash of frame data
                frame_hash = hashlib.sha256(frame.tobytes()).hexdigest()

                # Run YOLOv8 inference on this camera's frame
                results = self.model(frame, conf=MODEL_CONFIDENCE, verbose=False)
                detections = []

                # Process results for this camera
                for result in results:
                    boxes = result.boxes
                    for box in boxes:
                        class_id = int(box.cls[0])
                        confidence = float(box.conf[0])
                        bbox = box.xyxy[0].tolist()

                        # Only process relevant classes
                        if class_id in OBJECT_CLASSES:
                            object_type = OBJECT_CLASSES[class_id]
                            position = self.calculate_position(bbox, frame.shape[1], frame.shape[0], zone)

                            # Create detection with integrity data
                            detection_data = {
                                "object": object_type,
                                "position": position,
                                "confidence": confidence,
                                "bbox": bbox,
                                "class_id": class_id,
                                "camera_zone": zone,
                                "timestamp": time.time(),
                                "frame_hash": frame_hash[:16]  # Short hash for integrity
                            }

                            # Add HMAC for detection integrity (using a simple key for demo)
                            secret_key = os.environ.get('DETECTION_SECRET_KEY', 'default_key')
                            message = f"{object_type}{confidence}{zone}{time.time()}"
                            hmac_digest = hmac.new(secret_key.encode(), message.encode(), hashlib.sha256).hexdigest()
                            detection_data["integrity_hmac"] = hmac_digest[:16]

                            detections.append(detection_data)

                all_detections.extend(detections)

                # Send detections via Kafka with integrity checks
                if detections:
                    self.kafka_producer.send_detections(detections)

            except Exception as e:
                logger.error(f"❌ Error processing {zone} camera: {e}")

        return all_detections

    def draw_camera_status(self, frame: np.ndarray, camera_status: Dict) -> np.ndarray:
        """Draw camera status information on frame"""
        # Add camera status overlay
        status_text = "Camera Status: "
        for zone, info in camera_status.items():
            status_text += f"{zone}: {CAMERA_STATUS.get(info['status'], info['status'])} | "

        cv2.putText(frame, status_text[:-3], (10, 30),
                   cv2.FONT_HERSHEY_SIMPLEX, 0.6, (255, 255, 255), 2)

        return frame

    def play_alert_sound(self):
        """Play alert sound for blind spot detection"""
        try:
            self.alert_sound.play()
        except:
            pass  # Ignore audio errors

    def stop(self):
        """Stop all cameras and cleanup"""
        logger.info("🛑 Stopping multi-camera detection system...")
        self.is_running = False

        # Release all camera resources
        for zone, cap in self.cameras.items():
            if cap:
                cap.release()
                logger.info(f"📹 Released {zone} camera")

        self.cameras.clear()

        # Update status
        for zone in self.camera_status:
            self.camera_status[zone]["status"] = "not_connected"
            self.camera_status[zone]["cap"] = None

        # Close Kafka producer
        if self.kafka_producer:
            self.kafka_producer.stop_producer()
            logger.info("Kafka producer closed")

        cv2.destroyAllWindows()
        logger.info("✅ Multi-camera system stopped")


async def test_multi_camera_system():
    """Test function for the multi-camera system"""
    detector = MultiCameraDetector()

    try:
        # Start all cameras
        logger.info("🚀 Starting multi-camera test...")
        camera_results = detector.start_cameras()

        # Print camera status
        status = detector.get_camera_status()
        logger.info("📊 Camera Status:")
        for zone, info in status.items():
            logger.info(f"  {zone}: {info['name']} (ID: {info['camera_id']}) - {CAMERA_STATUS.get(info['status'], info['status'])}")

        # Process frames for a few seconds
        logger.info("🎥 Starting detection loop... (Press Ctrl+C to stop)")

        try:
            while True:  # Run continuously
                detections = await detector.process_all_cameras()

                # Calculate FPS
                detector.frame_count += 1
                current_time = time.time()
                if current_time - detector.last_time >= 1.0:
                    detector.fps = detector.frame_count / (current_time - detector.last_time)
                    detector.frame_count = 0
                    detector.last_time = current_time

                    # Log performance
                    logger.info(f"FPS: {detector.fps:.1f} | Active detections: {len(detections)}")

                # Play alert sound if objects in blind spots
                blind_spot_detections = []
                for detection in detections:
                    zone = detection["camera_zone"]
                    x_pos = detection["position"]["x"] / POSITION_SCALE["x"]
                    y_pos = detection["position"]["y"] / POSITION_SCALE["y"]

                    if detector.is_in_blind_spot(x_pos, y_pos, zone):
                        blind_spot_detections.append(detection)

                if blind_spot_detections:
                    detector.play_alert_sound()
                    logger.warning(f"🚨 BLIND SPOT ALERT! Objects detected: {len(blind_spot_detections)}")

                # Small delay to maintain target FPS
                await asyncio.sleep(1/FPS_TARGET)

        except KeyboardInterrupt:
            logger.info("⏹️  Test interrupted by user")
    except Exception as e:
        logger.error(f"❌ Test error: {e}")
    finally:
        detector.stop()


if __name__ == "__main__":
    asyncio.run(test_multi_camera_system())
